### PR TITLE
promoted validate commands to oadm

### PR DESCRIPTION
**Pull request by stevekuznetsov**
fixes https://github.com/openshift/origin/issues/6276

@deads2k PTAL

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
[test]

File: hack/test-cmd.sh [hunk 1]
```diff
@@ -133,16 +133,16 @@ openshift start \
   --images="${USE_IMAGES}"
 
 # validate config that was generated
-[ "$(openshift ex validate master-config ${MASTER_CONFIG_DIR}/master-config.yaml 2>&1 | grep SUCCESS)" ]
-[ "$(openshift ex validate node-config ${NODE_CONFIG_DIR}/node-config.yaml 2>&1 | grep SUCCESS)" ]
+[ "$(oadm validate master-config ${MASTER_CONFIG_DIR}/master-config.yaml 2>&1 | grep SUCCESS)" ]
```
Comment (deads2k) [inline hack/test-cmd.sh]: @jwforres @smarterclayton @fabianofranz How do you feel about this location?

**Comment by liggitt** [comment]
I'd also like these run as part of `openshift diagnostics` (can be done separately, but I didn't want to forget to mention it)

**Comment by liggitt** [comment]
actually, I wonder if these would be better as `openshift diagnostics -d NodeConfig`, etc

**Comment by stevekuznetsov** [comment]
Quite a lot of `go test` timed out and paniced as a result... [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/7912/consoleFull):
```
panic: test timed out after 1m0s
...
goroutine 7 [runnable]:
golang.org/x/crypto/blowfish.encryptBlock(0x8503eeb4e98b33c3, 0xc20809a000, 0x8503eeb4e98b33c3)
	/data/src/github.com/openshift/origin/Godeps/_workspace/src/golang.org/x/crypto/blowfish/block.go:115
...
FAIL	github.com/openshift/origin/pkg/auth/authenticator/password/htpasswd	60.234s
...
panic: test timed out after 1m0s
...
goroutine 20 [runnable]:
sort.(*StringSlice).Less(0xc20830cd60, 0xac, 0x0, 0x1)
	<autogenerated>:13
...
FAIL	github.com/openshift/origin/pkg/build/prune	60.531s
...
panic: test timed out after 1m0s
...
goroutine 41 [runnable]:
math/big.nat.norm(0xc2082465a0, 0x20, 0x24, 0xc20821d680, 0x10, 0x24)
	/usr/lib/golang/src/math/big/nat.go:61
...
FAIL	github.com/openshift/origin/pkg/cmd/server/crypto	60.451s
...
```

**Comment by liggitt** [comment]
I don't care about the panic... I care about the timeout

**Comment by stevekuznetsov** [comment]
@deads2k do we need to keep this in `openshift ex` as well?

**Comment by stevekuznetsov** [comment]
e2e flake [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8089/console):
```
!!! Error in hack/../test/end-to-end/core.sh:161
	'out=$(oc run cli-with-token-2 --attach --env=POD_NAMESPACE=test --image=openshift/origin:${TAG} --restart=Never -- cli whoami --loglevel=4 2>&1)' exited with status 1
Call stack:
	1: hack/../test/end-to-end/core.sh:161 main(...)
Exiting with status 1
```

#6149

File: hack/test-cmd.sh (hunk 1 above)
Comment (stevekuznetsov) [inline hack/test-cmd.sh]: @jwforres @smarterclayton @fabianofranz ping

Comment (fabianofranz) [inline hack/test-cmd.sh]: I'm not sure it should be a top-level command. Maybe we should also promote `diagnostics` and make it under `oadm diagnostics validate`. Also what's the difference of these `validate` commands and `openshift ex diagnostics --master-config=''`?

Comment (deads2k) [inline hack/test-cmd.sh]: > Also what's the difference of these validate commands and openshift ex diagnostics --master-config=''?

Diagnostics include a large number of heuristic tests in addition validating the configs themselves.

Comment (liggitt) [inline hack/test-cmd.sh]: the master config validation would be a single diagnostic that could be run...

Comment (deads2k) [inline hack/test-cmd.sh]: > the master config validation would be a single diagnostic that could be run...

@abutcher We'll probably do this and remove the `openshift ex` commands after `openshift ex diagnostics` has been promoted.

Comment (stevekuznetsov) [inline hack/test-cmd.sh]: Do we still want to expose these as first-level commands, or move them to `openshift diagnostics -d NodeConfig`, etc as @liggitt mentioned?

Comment (stevekuznetsov) [inline hack/test-cmd.sh]: @liggitt @deads2k @fabianofranz @jwforres @smarterclayton Final thoughts on where to put these commands?

**Comment by stevekuznetsov** [comment]
@smarterclayton OK, wasn't sure if that was always the case or only for the LDAP stuff since it was documented and published. TFTI

File: hack/test-cmd.sh (hunk 1 above)
Comment (deads2k) [inline hack/test-cmd.sh]: > @liggitt @deads2k @fabianofranz @jwforres @smarterclayton Final thoughts on where to put these commands?

Figure out how to express it using `openshift ex diagnostics` in a reasonable way.  We'll provide it via that mechanism.

**Comment by stevekuznetsov** [comment]
Added `ValidateMasterConfig` and `ValidateNodeConfig` as targets for cluster diagnostics. Kept the legacy code for validation so we can support the `openshift ex` endpoint. @sosiouxme please review diagnostics piece. @liggitt @smarterclayton PTAL, esp. about `openshift ex`.

File: pkg/cmd/experimental/diagnostics/cluster.go [hunk 2]
```diff
@@ -57,6 +57,10 @@ func (o DiagnosticsOptions) buildClusterDiagnostics(rawConfig *clientcmdapi.Conf
 			diagnostics = append(diagnostics, &clustdiags.ClusterRoles{ClusterRolesClient: clusterClient, SARClient: clusterClient})
 		case clustdiags.ClusterRoleBindingsName:
 			diagnostics = append(diagnostics, &clustdiags.ClusterRoleBindings{ClusterRoleBindingsClient: clusterClient, SARClient: clusterClient})
+		case clustdiags.ValidateMasterConfigName:
+			diagnostics = append(diagnostics, &clustdiags.ValidateMasterConfig{MasterConfigFile: o.MasterConfigLocation})
+		case clustdiags.ValidateNodeConfigName:
+			diagnostics = append(diagnostics, &clustdiags.ValidateNodeConfig{NodeConfigFile: o.NodeConfigLocation})
```
Comment (sosiouxme) [inline pkg/cmd/experimental/diagnostics/cluster.go]: Don't think these additions are necessary; current host diagnostics do master and node config validation already.

**Comment by stevekuznetsov** [comment]
@deads2k marked commands as deprecated

**Comment by deads2k** [comment]
There's a `Deprecated` field on the `Command` struct:  use that.

**Comment by stevekuznetsov** [comment]
@deads2k updated

**Comment by deads2k** [comment]
This will make that command more palatable I think: https://github.com/openshift/origin/pull/6634

**Comment by stevekuznetsov** [comment]
closing in favor of better deprecation messages following https://github.com/openshift/origin/pull/6634

**Pull request closed without merging**